### PR TITLE
Use 	DNS.GetHostAddressesAsync instead of sync version

### LINKED ISSUE #13163: Task: Use async DNS resolution in Socket and web stack code.
https://github.com/mono/mono/blob/b936c531018a354c92103cc9b7cf401231063725/mcs/class/System/System.Net.Sockets/Socket.cs#L975

It causes that `BeginConnect()` doesn't return immediately but returns until the name resolving completes, then unexpected blocking will happen in some async code calling this method.

### PR DESCRIPTION
**Pull request by MaximLipnin**
Fixes #13163

### CODE REVIEW

File: mcs/class/System/System.Net.Sockets/Socket.cs [hunk 1]
```diff
@@ -972,7 +972,9 @@ public IAsyncResult BeginConnect (string host, int port, AsyncCallback requestCa
 			if (is_listening)
 				throw new InvalidOperationException ();
 
-			return BeginConnect (Dns.GetHostAddresses (host), port, requestCallback, state);
+			IPAddress[] hostAddresses = Dns.GetHostAddressesAsync (host).GetAwaiter ().GetResult ();
```
Comment (filipnavara) [inline mcs/class/System/System.Net.Sockets/Socket.cs]: Unfortunately this doesn't really solve the problem. The `GetResult ()` call makes is synchronous again.

Comment (MaximLipnin) [reply mcs/class/System/System.Net.Sockets/Socket.cs]: Thanks for help!

File: mcs/class/System/System.Net.Sockets/Socket.cs [hunk 2]
```diff
@@ -972,7 +973,32 @@ public IAsyncResult BeginConnect (string host, int port, AsyncCallback requestCa
 			if (is_listening)
 				throw new InvalidOperationException ();
 
-			return BeginConnect (Dns.GetHostAddresses (host), port, requestCallback, state);
+			var callback = new AsyncCallback ((result) => {
+				IPAddress[] hostAddresses = ((Task<IPAddress[]>)result).Result;
+				BeginConnect (hostAddresses, port, requestCallback, state);
```
Comment (Therzok) [inline mcs/class/System/System.Net.Sockets/Socket.cs]: to avoid capture, maybe just use `((Task<IPAddress[]>)result).AsyncState`

Comment (marek-safar) [reply mcs/class/System/System.Net.Sockets/Socket.cs]: btw, this will still capture the context due to port and requestCallback parameters

**Pull request merged**